### PR TITLE
Support string type index to add | delete | update | scan

### PR DESCRIPTION
**Pull request by bright-starry-sky**
depend on #156

### CODE REVIEW

File: src/utils/IndexKeyUtils.h [hunk 1]
```diff
@@ -370,9 +382,8 @@ class IndexKeyUtils final {
                     break;
                 }
                 case Value::Type::STRING: {
-                    auto off = key.size() - vCount * sizeof(int32_t) - tailLen;
-                    len = *reinterpret_cast<const int32_t*>(key.c_str() + off);
-                    --vCount;
+                    len = (hasNullableCol && nullableBit.test(nullableColPosit))
```
Comment (critical27) [inline src/utils/IndexKeyUtils.h:385]: Why len is 1 when value is null?

Comment (bright-starry-sky) [reply src/utils/IndexKeyUtils.h:385]: > Why len is 1 when value is null?

If the value type is string and it is null. only one byte is filled.

**Comment by bright-starry-sky** [comment]
Resolved conflict; adressed comments

File: src/storage/exec/IndexScanNode.h [hunk 2]
```diff
@@ -43,7 +43,11 @@ class IndexScanNode : public RelNode<T> {
         if (ret != kvstore::ResultCode::SUCCEEDED) {
             return ret;
         }
-        scanPair_ = scanStr(partId);
+        auto scanRet = scanStr(partId);
+        if (!scanRet.ok()) {
+            return kvstore::ResultCode::ERR_CONSENSUS_ERROR;
```
[the remark this answers is not in the document]
Comment (bright-starry-sky) [reply src/storage/exec/IndexScanNode.h]: > Maybe use another errorcode?

Good point. agreed！

File: src/utils/IndexKeyUtils.h [hunk 3]
```diff
@@ -95,6 +95,20 @@ class IndexKeyUtils final {
         return raw;
     }
 
+    static std::string encodeValue(const Value& v, int16_t len) {
+        if (v.type() == Value::Type::STRING) {
+            std::string fs = v.getStr();
+                if (static_cast<size_t>(len) > v.getStr().size()) {
```
[the remark this answers is not in the document]
Comment (bright-starry-sky) [reply src/utils/IndexKeyUtils.h]: Thanks.  fixed.

**Review by critical27 — APPROVED** [review]
LGTM

File: src/utils/IndexKeyUtils.cpp [hunk 4]
```diff
@@ -144,7 +128,25 @@ IndexKeyUtils::collectIndexValues(RowReader* reader,
...
+            auto len = static_cast<size_t>(*col.type.get_type_length());
+            if (len > v.getStr().size()) {
+                fs.append(len - v.getStr().size(), '\0');
+            } else {
+                fs = fs.substr(0, len);
+            }
+            values.emplace_back(Value(fs));
+        } else {
+            values.emplace_back(std::move(v));
+        }
```
Comment (panda-sheep) [inline src/utils/IndexKeyUtils.cpp:149]: A small question:
If the result of `checkValue `  is ok,   just judge whether the result of v.isNull() is null. WDYT?

**Review by panda-sheep — COMMENTED** [review]
Good job!

File: src/utils/IndexKeyUtils.cpp (hunk 4 above)
Comment (bright-starry-sky) [reply src/utils/IndexKeyUtils.cpp:149]: > A small question:
> If the result of `checkValue ` is ok, just judge whether the result of`v.isNull()` is true or false. WDYT?

Good point. you are right.

Comment (bright-starry-sky) [reply src/utils/IndexKeyUtils.cpp:142]: improved.

**Review by panda-sheep — APPROVED** [review]
Excellent！

**Review by critical27 — APPROVED** [review]

**Pull request merged**